File: python/agentsentinel/inspector.py
# AgentSentinel — Safety controls for AI agents
# Copyright (c) 2026 Leland E. Doss. All rights reserved.
# Licensed under the Business Source License 1.1
# See LICENSE.md for details

"""Content inspection for outbound data."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional
from enum import Enum

from .pii import PIIScanner, PIIConfig, PIIMatch


class InspectionResult(Enum):
    """Result of content inspection."""
    ALLOW = "allow"
    BLOCK = "block"
    REDACT = "redact"
    WARN = "warn"


@dataclass
class InspectionReport:
    """Report from content inspection."""
    result: InspectionResult
    reason: str
    pii_matches: List[PIIMatch] = field(default_factory=list)
    redacted_content: Optional[str] = None
    metadata: Dict[str, Any] = field(default_factory=dict)


@dataclass
class InspectorConfig:
    """Configuration for content inspection.

    Parameters
    ----------
    enabled:
        Master switch for content inspection.
    pii_config:
        PII detection configuration.
    max_content_size:
        Maximum content size to inspect (larger content is blocked).
    inspect_tool_args:
        Whether to inspect tool arguments before execution.
    inspect_tool_results:
        Whether to inspect tool return values.
    block_on_pii:
        Block the request if PII is detected.
    custom_inspectors:
        Additional inspection functions: (content, tool_name) -> InspectionReport
    sensitive_data_threshold:
        Number of PII matches that triggers blocking (if block_on_pii=False).
    """
    enabled: bool = True
    pii_config: PIIConfig = field(default_factory=PIIConfig)
    max_content_size: int = 10_000_000  # 10MB
    inspect_tool_args: bool = True
    inspect_tool_results: bool = True
    block_on_pii: bool = True
    custom_inspectors: List[Callable[[Any, str], InspectionReport]] = field(default_factory=list)
    sensitive_data_threshold: int = 1


class ContentInspector:
    """Inspects content for sensitive data before allowing operations."""

    def __init__(self, config: InspectorConfig):
        self.config = config
        self.pii_scanner = PIIScanner(config.pii_config)

    def inspect(self, content: Any, tool_name: str, direction: str = "outbound") -> InspectionReport:
        """Inspect content for sensitive data.

        Parameters
        ----------
        content:
            The content to inspect (str, dict, or nested structure).
        tool_name:
            Name of the tool being used.
        direction:
            "outbound" for data leaving the system, "inbound" for responses.

        Returns
        -------
        InspectionReport
            The inspection result with details.
        """
        if not self.config.enabled:
            return InspectionReport(
                result=InspectionResult.ALLOW,
                reason="Inspection disabled"
            )

        # Check content size
        content_str = str(content)
        if len(content_str) > self.config.max_content_size:
            return InspectionReport(
                result=InspectionResult.BLOCK,
                reason=f"Content size {len(content_str)} exceeds limit {self.config.max_content_size}"
            )

        # Scan for PII
        pii_matches = self.pii_scanner.scan(content)

        if pii_matches:
            if self.config.block_on_pii or len(pii_matches) >= self.config.sensitive_data_threshold:
                return InspectionReport(
                    result=InspectionResult.BLOCK,
                    reason=f"PII detected: {len(pii_matches)} matches ({', '.join(m.pii_type.value for m in pii_matches[:3])}...)",
                    pii_matches=pii_matches,
                )
            else:
                # Redact and allow
                redacted = self.pii_scanner.redact(content_str)
                return InspectionReport(
                    result=InspectionResult.REDACT,
                    reason=f"PII redacted: {len(pii_matches)} matches",
                    pii_matches=pii_matches,
                    redacted_content=redacted,
                )

        # Run custom inspectors
        for inspector_fn in self.config.custom_inspectors:
            report = inspector_fn(content, tool_name)
            if report.result in (InspectionResult.BLOCK, InspectionResult.REDACT):
                return report

        return InspectionReport(
            result=InspectionResult.ALLOW,
            reason="Content inspection passed"
        )

    def inspect_args(self, tool_name: str, args: tuple, kwargs: dict) -> InspectionReport:
        """Inspect tool arguments before execution."""
        if not self.config.inspect_tool_args:
            return InspectionReport(result=InspectionResult.ALLOW, reason="Arg inspection disabled")

        combined = {"args": args, "kwargs": kwargs}
        return self.inspect(combined, tool_name, direction="outbound")

    def inspect_result(self, tool_name: str, result: Any) -> InspectionReport:
        """Inspect tool result after execution."""
        if not self.config.inspect_tool_results:
            return InspectionReport(result=InspectionResult.ALLOW, reason="Result inspection disabled")

        return self.inspect(result, tool_name, direction="inbound")

File: typescript/src/inspector.ts
/**
 * Content inspection for outbound data in AgentSentinel.
 */

import { PIIConfig, PIIMatch, PIIScanner } from "./pii";

/** Result of a content inspection. */
export enum InspectionResult {
  ALLOW = "allow",
  BLOCK = "block",
  REDACT = "redact",
  WARN = "warn",
}

/** Detailed report from a content inspection. */
export interface InspectionReport {
  result: InspectionResult;
  reason: string;
  piiMatches: PIIMatch[];
  redactedContent?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  metadata: Record<string, any>;
}

/** Custom inspector function signature. */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type InspectorFn = (content: any, toolName: string) => InspectionReport;

/** Configuration options for InspectorConfig. */
export interface InspectorConfigOptions {
  /** Master switch for content inspection. Default: true. */
  enabled?: boolean;
  /** PII detection configuration. */
  piiConfig?: PIIConfig;
  /** Maximum content size to inspect; larger content is blocked. Default: 10 MB. */
  maxContentSize?: number;
  /** Whether to inspect tool arguments before execution. Default: true. */
  inspectToolArgs?: boolean;
  /** Whether to inspect tool return values after execution. Default: true. */
  inspectToolResults?: boolean;
  /** Block the request if PII is detected. Default: true. */
  blockOnPII?: boolean;
  /** Additional inspection functions. */
  customInspectors?: InspectorFn[];
  /** Number of PII matches that triggers blocking when `blockOnPII` is false. Default: 1. */
  sensitiveDataThreshold?: number;
}

/** Immutable content inspector configuration value object. */
export class InspectorConfig {
  readonly enabled: boolean;
  readonly piiConfig: PIIConfig;
  readonly maxContentSize: number;
  readonly inspectToolArgs: boolean;
  readonly inspectToolResults: boolean;
  readonly blockOnPII: boolean;
  readonly customInspectors: readonly InspectorFn[];
  readonly sensitiveDataThreshold: number;

  constructor(options: InspectorConfigOptions = {}) {
    this.enabled = options.enabled ?? true;
    this.piiConfig = options.piiConfig ?? new PIIConfig();
    this.maxContentSize = options.maxContentSize ?? 10_000_000;
    this.inspectToolArgs = options.inspectToolArgs ?? true;
    this.inspectToolResults = options.inspectToolResults ?? true;
    this.blockOnPII = options.blockOnPII ?? true;
    this.customInspectors = Object.freeze(options.customInspectors ?? []);
    this.sensitiveDataThreshold = options.sensitiveDataThreshold ?? 1;
  }
}

/** Inspects content for sensitive data before allowing operations. */
export class ContentInspector {
  private readonly config: InspectorConfig;
  private readonly piiScanner: PIIScanner;

  constructor(config: InspectorConfig) {
    this.config = config;
    this.piiScanner = new PIIScanner(config.piiConfig);
  }

  /**
   * Inspect content for sensitive data.
   *
   * @param content   The content to inspect (string, object, or nested structure).
   * @param toolName  Name of the tool being used (for reporting).
   * @param direction `"outbound"` for data leaving the system, `"inbound"` for responses.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  inspect(content: any, toolName: string, direction: "outbound" | "inbound" = "outbound"): InspectionReport {
    if (!this.config.enabled) {
      return { result: InspectionResult.ALLOW, reason: "Inspection disabled", piiMatches: [], metadata: {} };
    }

    const contentStr = String(content);
    if (contentStr.length > this.config.maxContentSize) {
      return {
        result: InspectionResult.BLOCK,
        reason: `Content size ${contentStr.length} exceeds limit ${this.config.maxContentSize}`,
        piiMatches: [],
        metadata: { direction },
      };
    }

    const piiMatches = this.piiScanner.scan(content);

    if (piiMatches.length > 0) {
      if (this.config.blockOnPII || piiMatches.length >= this.config.sensitiveDataThreshold) {
        const topTypes = piiMatches.slice(0, 3).map(m => m.piiType).join(", ");
        return {
          result: InspectionResult.BLOCK,
          reason: `PII detected: ${piiMatches.length} matches (${topTypes}...)`,
          piiMatches,
          metadata: { direction },
        };
      } else {
        const redacted = this.piiScanner.redact(contentStr);
        return {
          result: InspectionResult.REDACT,
          reason: `PII redacted: ${piiMatches.length} matches`,
          piiMatches,
          redactedContent: redacted,
          metadata: { direction },
        };
      }
    }

    // Run custom inspectors
    for (const inspectorFn of this.config.customInspectors) {
      const report = inspectorFn(content, toolName);
      if (report.result === InspectionResult.BLOCK || report.result === InspectionResult.REDACT) {
        return report;
      }
    }

    return { result: InspectionResult.ALLOW, reason: "Content inspection passed", piiMatches: [], metadata: { direction } };
  }

  /** Inspect tool arguments before execution. */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  inspectArgs(toolName: string, args: any[]): InspectionReport {
    if (!this.config.inspectToolArgs) {
      return { result: InspectionResult.ALLOW, reason: "Arg inspection disabled", piiMatches: [], metadata: {} };
    }
    return this.inspect({ args }, toolName, "outbound");
  }

  /** Inspect a tool result after execution. */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  inspectResult(toolName: string, result: any): InspectionReport {
    if (!this.config.inspectToolResults) {
      return { result: InspectionResult.ALLOW, reason: "Result inspection disabled", piiMatches: [], metadata: {} };
    }
    return this.inspect(result, toolName, "inbound");
  }
}
